src/matrixrtc/CallMembership.ts

```diff
@@ -78,31 +134,39 @@ export class CallMembership {
         return this.data.application;
     }
 
-    public get scope(): CallScope {
+    public get scope(): CallScope | undefined {
         return this.data.scope;
     }
 
     public get membershipID(): string {
-        return this.data.membershipID;
+        if (isLegacyCallMembershipData(this.data)) return this.data.membershipID;
+        // the createdTs behaves equivalent to the membershipID.
+        // we only need the field for the legacy member envents where we needed to update them
+        // synapse ignores sending state events if they have the same content.
+        else return this.createdTs().toString();
     }
 
     public createdTs(): number {
         return this.data.created_ts ?? this.parentEvent.getTs();
     }
 
-    public getAbsoluteExpiry(): number {
+    public getAbsoluteExpiry(): number | undefined {
+        if (!isLegacyCallMembershipData(this.data)) return undefined;
         if (this.data.expires) {
-            return this.createdTs() + this.data.expires;
+            // we know createdTs exists since we already do the isLegacyCallMembershipData check
+            return this.createdTs()! + this.data.expires;
         } else {
             // We know it exists because we checked for this in the constructor.
             return this.data.expires_ts!;
         }
     }
```

Comment:
If `CallMembershipData` were an `XOR` type, the availability of fields could be deduced by static typechecking & the non-null assertion operators here could be removed.

Reply:
I dont see how we could remove the non null assertion since expires_ts is an optional field on CallMembershipLegacyData and a missing field on SessionMembershipData.